File: python/dlq-to-s3.py
import boto3
from botocore.exceptions import NoCredentialsError

# Initialize Boto3 clients for SQS and S3
sqs_client = boto3.client('sqs')
s3_client = boto3.client('s3')

# AWS SQS and S3 configuration
dead_letter_queue_url = 'your-dlq-url'
bucket_name = 'your-bucket-name'
file_path = 'your-file-path'

def fetch_messages_from_dlq():
    """
    Fetch messages from the Dead Letter Queue.
    """
    try:
        messages = sqs_client.receive_message(
            QueueUrl=dead_letter_queue_url,
            MaxNumberOfMessages=10,  # Adjust based on your needs
            WaitTimeSeconds=20  # Long polling
        )
        return messages.get('Messages', [])
    except NoCredentialsError:
        print("Credentials not available")
        return []

def save_messages_to_s3(messages):
    """
    Save messages to an S3 bucket.
    """
    if not messages:
        print("No messages to save")
        return

    # Convert message list to string, one message per line
    message_content = '\n'.join([message['Body'] for message in messages])
    
    try:
        # Check if the file already exists in S3 and append content if it does
        try:
            existing_content = s3_client.get_object(Bucket=bucket_name, Key=file_path)
            existing_content_body = existing_content['Body'].read().decode('utf-8')
            message_content = existing_content_body + '\n' + message_content
        except s3_client.exceptions.NoSuchKey:
            print("File not found in S3. Creating new file.")
        
        # Upload the new content
        s3_client.put_object(Bucket=bucket_name, Key=file_path, Body=message_content)
        print("Successfully uploaded messages to S3")
    except NoCredentialsError:
        print("Credentials not available")
    except Exception as e:
        print(f"Failed to upload messages to S3: {e}")

def delete_messages_from_dlq(messages):
    """
    Delete messages from the DLQ once they have been successfully saved to S3.
    """
    for message in messages:
        try:
            sqs_client.delete_message(
                QueueUrl=dead_letter_queue_url,
                ReceiptHandle=message['ReceiptHandle']
            )
            print(f"Message {message['MessageId']} deleted from DLQ")
        except Exception as e:
            print(f"Failed to delete message {message['MessageId']} from DLQ: {e}")

if __name__ == "__main__":
    messages = fetch_messages_from_dlq()
    if messages:
        save_messages_to_s3(messages)
        delete_messages_from_dlq(messages)
    else:
        print("No messages found in DLQ.")

File: nodejs/dlq-to-s3.js
/**
 * Script to transfer messages from an AWS SQS Dead Letter Queue (DLQ) to an S3 bucket.
 * This prevents messages from expiring in the DLQ by storing them for later processing.
 */

import { SQSClient, ReceiveMessageCommand, DeleteMessageCommand } from '@aws-sdk/client-sqs';
import { GetObjectCommand, S3Client, PutObjectCommand } from '@aws-sdk/client-s3';
import { fromIni } from '@aws-sdk/credential-provider-ini';

// AWS configuration
const region = "your-region"; // Example: "us-east-1"
const credentials = fromIni({ profile: 'your-profile' }); // Profile in your AWS credentials file

// SQS and S3 configuration
const s3Client = new S3Client({ region, credentials });
const sqsClient = new SQSClient({ region, credentials });

// Replace these with your Dead Letter Queue URL and S3 bucket details
const bucketName = 'your-bucket-name'; // Example: 'your-dlq-messages-bucket'
const deadLetterQueueURL = 'your-dlq-url'; // Example: 'https://sqs.your-region.amazonaws.com/your-account-id/your-dlq-name'
const filePath = 'your-file-path'; // Example: 'dlq-messages.txt'

// Receive messages from the DLQ
const receiveMessages = async () => {
  const command = new ReceiveMessageCommand({
    QueueUrl: deadLetterQueueURL,
    MaxNumberOfMessages: 10,
    WaitTimeSeconds: 20
  });

  try {
    const data = await sqsClient.send(command);
    return data.Messages || [];
  } catch (error) {
    console.error('Error receiving messages:', error);
    throw error;
  }
};

// Get existing content from S3 (if any)
const getExistingS3Content = async () => {
  try {
    const data = await s3Client.send(new GetObjectCommand({
      Bucket: bucketName,
      Key: filePath,
    }));
    const bodyContents = await streamToString(data.Body);
    return bodyContents;
  } catch (error) {
    if (error.name === 'NoSuchKey') {
      return ''; // Return empty string if the file does not exist
    } else {
      console.error('Error downloading from S3:', error);
      throw error;
    }
  }
};

// Helper function to convert stream to string
function streamToString(stream) {
  return new Promise((resolve, reject) => {
    const chunks = [];
    stream.on('data', (chunk) => chunks.push(chunk));
    stream.on('error', reject);
    stream.on('end', () => resolve(Buffer.concat(chunks).toString('utf-8')));
  });
}

// Write messages to S3
const writeToS3 = async (messages) => {
  const existingContent = await getExistingS3Content();
  const newContent = messages.map(msg => msg.Body).join('\n');
  const fileContent = existingContent + '\n' + newContent;
  const command = new PutObjectCommand({
    Bucket: bucketName,
    Key: filePath,
    Body: fileContent
  });

  try {
    await s3Client.send(command);
    console.log('Successfully uploaded messages to S3');
  } catch (error) {
    console.error('Error uploading to S3:', error);
    throw error;
  }
};

// Delete a message from the DLQ
const deleteMessage = async (receiptHandle) => {
  const deleteParams = {
    QueueUrl: deadLetterQueueURL,
    ReceiptHandle: receiptHandle,
  };

  try {
    await sqsClient.send(new DeleteMessageCommand(deleteParams));
    console.log('Message deleted successfully');
  } catch (error) {
    console.error('Error deleting message:', error);
  }
};

const processDLQMessages = async () => {
  const messages = await receiveMessages();
  if (messages.length > 0) {
    await writeToS3(messages);
    for (const message of messages) {
      await deleteMessage(message.ReceiptHandle); // Delete after writing to S3
    }
    processDLQMessages(); // Recursively process next batch
  } else {
    console.log('No messages to process');
  }
};

processDLQMessages(); // Start the process
